### PR TITLE
Ported immutable media impls from scenegraph.

### PR DESCRIPTION
**Pull request by grokys**
The deferred renderer requires immutable objects for drawing as it renders on a separate thread. This PR makes `IGeometryImpl`, `IFormattedTextImpl` etc immutable so that they can be used in this manner.

`FormattedText` was previously a bit of a strange beast in that it was kind of immutable in that its `Text` couldn't be changed, this PR changes it so that it is fully mutable, backed with an immutable impl which is created on demand.

### CODE REVIEW

File: src/Avalonia.Visuals/Media/FormattedText.cs [hunk 1]
```diff
@@ -10,128 +10,109 @@ namespace Avalonia.Media
     /// <summary>
     /// Represents a piece of text with formatting.
     /// </summary>
-    public class FormattedText : AvaloniaDisposable
+    public class FormattedText
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Media/FormattedText.cs]: Why is it not disposable anymore? At least for Direct2D and Cairo underlying implementations are.

File: src/Gtk/Avalonia.Cairo/CairoPlatform.cs [hunk 2]
```diff
@@ -42,14 +42,20 @@ public IBitmapImpl CreateBitmap(int width, int height)
...
-            string fontFamily,
-            double fontSize,
-            FontStyle fontStyle,
+            Typeface typeface,
             TextAlignment textAlignment,
-            Avalonia.Media.FontWeight fontWeight,
-            TextWrapping wrapping)
+            TextWrapping wrapping,
+            Size constraint,
+            IReadOnlyList<FormattedTextStyleSpan> spans)
```
Comment (kekekeks) [inline src/Gtk/Avalonia.Cairo/CairoPlatform.cs]: :thumbsup:

File: tests/Avalonia.Visuals.UnitTests/Media/TypefaceTests.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using Avalonia.Media;
+using Xunit;
+
+namespace Avalonia.Visuals.UnitTests.Media
+{
+    public class TypefaceTests
+    {
+        [Fact]
+        public void Exception_Should_Be_Thrown_If_FontSize_0()
```
Comment (kekekeks) [inline tests/Avalonia.Visuals.UnitTests/Media/TypefaceTests.cs]: It should be `<= 0`, I think

File: src/Avalonia.Visuals/Media/FormattedText.cs (hunk 1 above)
Comment (grokys) [inline src/Avalonia.Visuals/Media/FormattedText.cs]: Yeah, the problem is that the impl may still be needed for the next frame so can't be immediately disposed. I considered adding a "dispose when the render has finished" flag but that started looking like manual memory management, which I'd prefer to avoid if possible. We could always add it in future if necessary.

File: tests/Avalonia.Visuals.UnitTests/Media/TypefaceTests.cs (hunk 3 above)
Comment (grokys) [inline tests/Avalonia.Visuals.UnitTests/Media/TypefaceTests.cs]: 👍

File: src/Avalonia.Visuals/Media/FormattedText.cs (hunk 1 above)
Comment (kekekeks) [inline src/Avalonia.Visuals/Media/FormattedText.cs]: The problem is that GC is unable to correctly determine the amount of memory occupied by texts, bitmaps, geometries, etc. So we might need some kind of reference counting wrapper anyway.

Comment (grokys) [inline src/Avalonia.Visuals/Media/FormattedText.cs]: Well there is the [GC.AddMemoryPressure](https://msdn.microsoft.com/en-us/library/system.gc.addmemorypressure(v=vs.110).aspx) method, but I agree we'll probably want to be able to dispose of these objects at some point. I'd prefer to do that in another PR however, as I'd like to get the scenegraph stuff ported over this year ;)
